File: member-service/src/main/java/com/hedvig/memberservice/query/MemberRepository.java
package com.hedvig.memberservice.query;

import java.util.List;
import java.util.Optional;
import java.util.stream.Stream;

import com.hedvig.memberservice.aggregates.MemberStatus;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;

public interface MemberRepository extends JpaRepository<MemberEntity, Long> {

  Optional<MemberEntity> findBySsn(String s);

  List<MemberEntity> findAllByIdIn(List<Long> ids);

  @Query("select count(*) from MemberEntity m where m.status = 'SIGNED'")
  Long countSignedMembers();

  @Query("SELECT m FROM MemberEntity m")
  Page<MemberEntity> searchAll(Pageable p);

  @Query(
      "SELECT m FROM MemberEntity m WHERE lower(m.firstName) LIKE lower(concat('%', :query, '%')) "
          + "OR lower(m.lastName) LIKE lower(concat('%', :query, '%'))")
  Page<MemberEntity> searchByQuery(@Param("query") String query, Pageable p);

  @Query("SELECT m FROM MemberEntity m WHERE status = :status")
  Page<MemberEntity> searchByStatus(@Param("status") MemberStatus status, Pageable p);

  @Query(
      "SELECT m FROM MemberEntity m WHERE status = :status "
          + "AND (lower(m.firstName) LIKE lower(concat('%', :query, '%')) "
          + "OR lower(m.lastName) LIKE lower(concat('%', :query, '%')))")
  Page<MemberEntity> searchByStatusAndQuery(
      @Param("status") MemberStatus status, @Param("query") String query, Pageable p);

  @Query("select m from MemberEntity m where m.id = :id")
  Page<MemberEntity> searchById(@Param("id") Long id, Pageable p);

  @Query("select m from MemberEntity m where m.id = :id and m.status = :status")
  Page<MemberEntity> searchByIdAndStatus(@Param("id") Long id, @Param("status") MemberStatus status, Pageable p);
}

File: member-service/src/main/java/com/hedvig/memberservice/query/MemberRepository.kt
package com.hedvig.memberservice.query

import com.hedvig.memberservice.aggregates.MemberStatus
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Query
import org.springframework.data.repository.query.Param

interface MemberRepository : JpaRepository<MemberEntity, Long> {

    fun findBySsn(ssn: String): List<MemberEntity>

    fun findAllByIdIn(ids: List<Long>): List<MemberEntity>

    @Query(
        """
      FROM MemberEntity
      WHERE status != 'SIGNED' AND (ssn = :ssn OR email = :email) AND id != :memberId
    """
    )
    fun findNonSignedBySsnOrEmailAndNotId(
        ssn: String,
        email: String,
        memberId: Long
    ): List<MemberEntity>

    @Query(
        """
        FROM MemberEntity
        WHERE (status = 'SIGNED' OR status='TERMINATED') AND (ssn = :ssn OR email = :email)
    """
    )
    fun findSignedMembersBySsnOrEmail(
        ssn: String?,
        email: String
    ): List<MemberEntity>

    @Query("select count(*) from MemberEntity m where m.status = 'SIGNED'")
    fun countSignedMembers(): Long?

    @Query(
        """
        SELECT m
        FROM MemberEntity m
        WHERE
            m.status IN ('SIGNED', 'TERMINATED')
        AND
            (
                CAST(m.id as text) = :query
                OR LOWER(TRIM(m.firstName)) LIKE ('%' || LOWER(:query) || '%')
                OR LOWER(TRIM(m.lastName)) LIKE ('%' || LOWER(:query) || '%')
                OR LOWER(TRIM(m.firstName) || ' ' || TRIM(m.lastName)) LIKE ('%' || LOWER(:query) || '%')
                OR m.ssn LIKE ('%' || :query || '%')
                OR LOWER(m.email) LIKE ('%' || LOWER(:query) || '%')
                OR m.phoneNumber LIKE ('%' || :query || '%')
            )
    """
    )
    fun searchSignedOrTerminated(
        @Param("query") query: String,
        p: Pageable
    ): Page<MemberEntity>

    @Query(
        """
        SELECT m
        FROM MemberEntity m
        WHERE
            CAST(m.id as text) = :query
            OR LOWER(TRIM(m.firstName)) LIKE ('%' || LOWER(:query) || '%')
            OR LOWER(TRIM(m.lastName)) LIKE ('%' || LOWER(:query) || '%')
            OR LOWER(TRIM(m.firstName) || ' ' || TRIM(m.lastName)) LIKE ('%' || LOWER(:query) || '%')
            OR m.ssn LIKE ('%' || :query || '%')
            OR LOWER(m.email) LIKE ('%' || LOWER(:query) || '%')
            OR m.phoneNumber LIKE ('%' || :query || '%')
    """
    )
    fun searchAll(
        @Param("query") query: String,
        p: Pageable
    ): Page<MemberEntity>

    fun findByStatus(status: MemberStatus): List<MemberEntity>

    fun findAllByStatusAndSsnNotIn(status: MemberStatus, ssns: List<String>): List<MemberEntity>

    @Query(
        """
        SELECT m.id from MemberEntity m WHERE m.pickedLocale is null and (first_name is null or first_name != 'GDPR')
    """
    )
    fun findIdsWithNoPickedLocale(pageable: Pageable): Page<Long>
}
